internal/subscriber/retry/retrier.go

```diff
@@ -0,0 +1,172 @@
+package retry
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/razorpay/metro/internal/brokerstore"
+	"github.com/razorpay/metro/internal/subscription"
+	"github.com/razorpay/metro/pkg/logger"
+	"github.com/razorpay/metro/pkg/messagebroker"
+)
+
+const (
+	defaultBrokerOperationsTimeoutMs int64 = 100
+)
+
+// IRetrier interface over retrier core functionalities.
+type IRetrier interface {
+	Handle(context.Context, messagebroker.ReceivedMessage) error
+	Stop(context.Context)
+}
+
+// Retrier implements all business logic for IRetrier
+type Retrier struct {
+	subs           *subscription.Model
+	bs             brokerstore.IBrokerStore
+	delayConsumers sync.Map
+}
+
+// NewRetrier inits a new retrier which internally takes care of spawning the needed delay-consumers.
+func NewRetrier(ctx context.Context, subs *subscription.Model, bs brokerstore.IBrokerStore, handler MessageHandler) (IRetrier, error) {
+	delayConsumers := sync.Map{}
+
+	for interval, topic := range subs.GetDelayTopicsMap() {
+		dc, err := NewDelayConsumer(ctx, topic, subs, bs, handler)
+		if err != nil {
+			return nil, err
+		}
+		go dc.Run(ctx)                     // run the delay consumer
+		delayConsumers.Store(interval, dc) // store the delay consumer for lookup
+	}
+
+	r := &Retrier{
+		subs:           subs,
+		bs:             bs,
+		delayConsumers: delayConsumers,
+	}
+
+	return r, nil
+}
+
+// Stop gracefully stop call the spawned delay-consumers for retry
+func (r *Retrier) Stop(ctx context.Context) {
+	logger.Ctx(ctx).Infow("retrier: stopping all delay consumers for topics", "delay_topics", r.subs.GetDelayTopics())
+	wg := sync.WaitGroup{}
+	r.delayConsumers.Range(func(_, dcFromMap interface{}) bool {
+		wg.Add(1)
+		dc := dcFromMap.(*DelayConsumer)
+		go func(wg *sync.WaitGroup) {
+			defer wg.Done()
+			dc.cancelFunc()
+			// wait for delay-consumer shutdown to complete
+			<-dc.doneCh
+		}(&wg)
+		return true
+	})
+	wg.Wait()
+}
+
+// Handle takes care of processing a given retry message.
+func (r *Retrier) Handle(ctx context.Context, msg messagebroker.ReceivedMessage) error {
+
+	logger.Ctx(ctx).Infow("retrier: received msg for retry", msg.LogFields()...)
+
+	availableDelayIntervals := subscription.Intervals
```

Comment:
move intervals constants under retry module ?

Reply:
I think this will cause a cyclic dependency.